README
/*!


\mainpage SDL_gfx - SDL graphics drawing primitives, rotozoom and other supporting functions


\section contact_sec Contact and License

Email aschiffler at ferzkopp.net to contact the author or better check
author's homepage at http://www.ferzkopp.net for the most up-to-date
contact information.

This library is licenced under the LGPL, see the file LICENSE for details. 

LGPL (c) A. Schiffler


\section intro_sec Introduction

The SDL_gfx library evolved out of the SDL_gfxPrimitives code which
provided basic drawing routines such as lines, circles or polygons for 
SDL Surfaces and adding a couple other useful functions for zooming 
images for example and doing basic image processing on byte arrays.

The current components of the SDL_gfx library are:
- Graphic Primitives (SDL_gfxPrimitives.h, SDL_gfxPrimitives.c)
- Rotozoomer (SDL_rotozoom.h, SDL_rotozoom.c)
- Framerate control (SDL_framerate.h, SDL_framerate.c)
- MMX image filters (SDL_imageFilter.h, SDL_imageFilter.c)
- Custom Blit functions (SDL_gfxBlitFunc.h, SDL_gfxBlitFunc.c)
- Build-in 8x8 Font (SDL_gfxPrimitives_font.h)



\subsection notes_gfx Notes on Graphics Primitives

Care has been taken so that all routines are fully alpha-aware and can 
blend any primitive onto the target surface if ALPHA<255. Surface depths 
supported are 1,2,3 and 4 bytes per pixel. Surface locking is implemented
in each routine and the library should work well with hardware 
accelerated surfaces. 

\htmlonly
<a href="../Screenshots/SDL_gfxPrimitives.jpg" target="_blank" title="SDL_gfxPrimitives Screenshot"><img src="../Screenshots/SDL_gfxPrimitives-thumb.jpg" border="0" hspace="5"></a><br />
\endhtmlonly

Currently, The following Anti-Aliased drawing primitives are available:
- AA-line
- AA-polygon
- AA-circle
- AA-ellipse

Note: All ___Color routines expect the color to be in the format 0xRRGGBBAA.

\subsection notes_roto Notes on Rotozoomer

The rotozoom code is not ASSEMBLY quality - but it should be fast enough 
even for some realtime effects if the CPU is good or bitmaps small.
With interpolation the routines are typically used for pre-rendering stuff 
in higher quality (i.e. smoothing) - that's also the reason why the API differs 
from SDL_BlitRect() - as they create a new target surface each time rotozoom 
is called. The final rendering speed is dependent on the target surface
size as it is beeing xy-scanned when rendering the new surface.

\htmlonly
<a href="../Screenshots/SDL_rotozoom.jpg" target="_blank" title="SDL_rotozoom Screenshot"><img src="../Screenshots/SDL_rotozoom-thumb.jpg" border="0" hspace="5"></a><br />
\endhtmlonly

Note also that the smoothing toggle is dependent on the input surface bit 
depth. 8bit surfaces will \b never be smoothed - only 32bit surfaces will.

Note that surfaces of other bit depth then 8 and 32 will be converted 
on the fly to a 32bit surface using a blit into a temporary surface. This 
impacts performance somewhat.

Smoothing (interpolation) flags work only on 32bit surfaces:
\verbatim
 #define SMOOTHING_OFF		0
 #define SMOOTHING_ON		1
\endverbatim
 
\subsection notes_rate Notes on Framerate Manager

The framerate functions are used to insert delays into the graphics loop
to maintain a constant framerate.

The implementation is more sophisticated that the usual
\verbatim
	SDL_Delay(1000/FPS); 
\endverbatim
call since these functions keep track of the desired game time per frame 
for a linearly interpolated sequence of future timing points of each frame. 
This is done to avoid rounding errors from the inherent instability in the 
delay generation and application.

\htmlonly
<a href="../framerate.png" target="_blank" title="Framerate Diagram"><img src="../framerate-thumb.png" border="0"></a><br />
\endhtmlonly

i.e. the 100th frame of a game running at 50Hz will be accurately
2.00sec after the 1st frame (if the machine can keep up with the drawing).

The functions return 0 or 'value' for sucess and -1 for error. All functions
use a pointer to a framerate-manager variable to operate.


\subsection notes_filter Notes on ImageFilters

The imagefilter functions are a collection of MMX optimized routines that
operate on continuous buffers of bytes - typically greyscale images from 
framegrabbers and such - performing functions such as image addition and 
binarization. All functions (almost ... not the the convolution routines) 
have a C implementation that is automatically used on systems without MMX 
capabilities.

The compiler flag -DUSE_MMX toggles the conditional compile of MMX assembly.
An assembler must be installed (i.e. "nasm").


\subsection platforms Supported Platforms

\subsubsection platformlinux Unix/Linux

The library compiles and is tested for a Linux target (gcc compiler) via the
the usual configure;make;make install sequence.

\subsubsection platformwindows Windows
 
A Win32 target is available (VisualC6/7/8/9, mingw32, xmingw32 cross-compiler).
The SDL_gfx.sln will open in VS2008 and VS2010 including express versions.
See "Other Builds" for additional makefiles (may be out of date).

When using the cross-compiler (available on the author's homepage, slightly
out of date), the build process generates .DLLs. You can use the command 
line 'LIB.EXE' tool to generate VC6 compatible .LIB files for linking 
purposes. 

\subsubsection platformosx Mac OSX 

Xcode is supported via templates. See "Other Builds" folder Xcode3+.zip -
this template only supports SDL_gfx and not the tests. For this template, the
Deployment Target (the lowest version to run on) is set to 10.5 and expects
the SDL.framework preinstalled in some default location
(either /Library/Frameworks, or ~/Library/Frameworks). Older targets are also reported to work (10.3+ native and Project Builder). 

\subsubsection platformqnx QNX

QNX is reported to build well (see .diff in "Other Builds").

\subsubsection platformzune Zune

ZuneHD (WinCE 6 ARM) is reported to build (see OpenZDK in "Other Builds").
Note that between rendering on the Zune's ARM CPU and constantly uploading
textures to the GPU, SDL_gfx is going to be slow. Also, the libc math 
functions all use software FP emulation even when VFP floating point
support is turned on in the compiler, so there's extra overhead due to that
as well.

\subsubsection platformothers Others

Other platforms might work but have not been tested by the author.
Please check the file "INSTALL" as well as the folder "Other Builds".

See also section "Installation" below for more build instructions.

\section install_sec Installation

\subsection unix Unix/Linux

To compile the library your need the SDL 1.2 installed from source or 
installed with the 'devel' RPM package. For example on Mandriva, run:
\verbatim
	urpmi  libSDL1.2-devel
\endverbatim

Then run
\verbatim
	./autogen.sh	# (optional)
	./configure
	make
	make install
	ldconfig
\endverbatim

to compile and install the library. The default location for the 
installation is /usr/local/lib and /usr/local/include. The libary 
path might need to be added to the file:
	/etc/ld.so.conf

Run the shell script 'nodebug.sh' before make, to patch the makefile 
for optimized compilation:
\verbatim
	./autogen.sh	# (optional)
	./configure
	./nodebug.sh
	make
	make install
	ldconfig
\endverbatim

Check the folder "Other Builds" for alternative makefiles.

\subsection prep Build Prep

Run autogen.sh or manually:
\verbatim
	aclocal --force
	libtoolize --force --copy
	autoreconf -fvi
\endverbatim

\subsection nommx No-MMX

To build without MMX code enabled (i.e. PPC or for AMD64 architecture
which is missing pusha/popa):
\verbatim
	./configure --disable-mmx
	make
	make install
\endverbatim
i.e. to build on MacOSX 10.3+ use:
\verbatim
	./configure --disable-mmx && make
\endverbatim

\subsection vs9 Windows (VC9)

Open solution file and review README.


\subsection vc6 Windows (VC6/7)

See folder Other Builds.

To create a Windows DLL using VisualC6:
\verbatim
	unzip -a VisualC6.zip
	vcvars32.bat
	copy VisualC/makefile
	nmake
\endverbatim
or
\verbatim
	unzip -a VisualC7.zip
\endverbatim
and open the project file.


\subsection wince WindowsCE

See folder Other Builds.

May need workaround for missing lrint.


\subsection cross Cross-Compilation

To build using mingw32 on Win32, check the makefile contained in mingw.zip

To create a Windows DLL using the xmingw32 cross-compiler:
\verbatim
	cross-configure
	cross-make
	cross-make install
\endverbatim

Make sure the -DBUILD_DLL is used (and only then) when creating the DLLs.
Make sure -DWIN32 is used when compiling the sources (creating or using
the DLLs.

Specify the path to your cross-compiled 'sdl-config', and invoke
'./configure' with the '--host' and '--build' arguments. For example,
to cross-compile a .DLL from GNU/Linux:
\verbatim
	SDL_CONFIG=/usr/local/cross-tools/i386-mingw32msvc/bin/sdl-config \
		./configure --host=i586-mingw32msvc --build=i686-pc-linux-gnu
	make
	make install
\endverbatim

\subsection qnx QNX

To build on QNX6, patch first using:
\verbatim
	patch -p0 <QNX.diff
\endverbatim

\subsection osx OSX

Use standard unix build sequence.

To build on MacOS X with Project Builder, follow these steps:
- Update your developer tools to the lastest version.
- Install the SDL Developers framework for Mac OS X.
- Download the latest SDL_gfx source distribution and extract the
	archive in a convenient location.
- Extract the included OSX-PB.tgz archive into the
	top directory of the SDL_gfx distribution (from step 3). This will create a
	PB that contains the project files.
- The project has targets for the SDL_gfx framework and the four test
	programs. All can be built using the 'deployment' or 'development' build
	styles. 

A newer version for MaxOS X is included in the OSX-PB-XCode.zip archive. The 
updated version uses relative pathnames where appropriate, and pointers to 
the standard  installation location of SDL. However, it may require XCode in 
order to be used.



\section test_sec Test Programs

Change to the ./Test directory and run
\verbatim
	./configure
	make
\endverbatim
to create several test programs for the libraries functions. This requires
the library to be compiled and installed.


See the source code .c files for some sample code and implementation hints.


\section contrib_sec Contributors

- Fix for filledbox by Ingo van Lil, inguin at gmx.de - thanks Ingo.

- Non-alpha line drawing code adapted from routine 
  by Pete Shinners, pete at shinners.org - thanks Pete.

- More fixes by Karl Bartel, karlb at gmx.net - thanks Karl.

- Much testing and suggestions for fixes from Danny van Bruggen,
  danny at froukepc.dhs.org - thanks Danny.

- Original AA-circle/-ellipse code idea from Stephane Magnenat, 
  nct at wg0.ysagoon.com - thanks Stephane.

- Faster blending routines contributed by Anders Lindstroem,
  cal at swipnet.se - thanks Anders.

- New AA-circle/-ellipse code based on ideas from Anders Lindstroem - 
  thanks Anders.

- VisualC makefile contributed by Danny van Bruggen, 
  danny at froukepc.dhs.org - thanks Danny.

- VisualC7 project file contributed by James Turk, 
  jturk at conceptofzero.com - thanks James.

- Project Builder package contributed by Thomas Tongue, 
  TTongue at imagiware.com - Thanks Thomas.

- Fix for filledPolygon contributed by Kentaro Fukuchi 
  fukuchi at is.titech.ac.jp - Thanks Kentaro.

- QNX6 patch contributed by Mike Gorchak,
  mike at malva.ua - Thanks Mike.

- Pie idea contributed by Eike Lange,
  eike.lange at uni-essen.de - Thanks Eike.

- Dynamic font setup by Todor Prokopov,
  koprok at dir.bg - Thanks Todor.

- Horizontal/Vertical flipping code by Victor (Haypo) 

ChangeLog
CHANGES/VERSION
===============

Ver 2.0.23 - in progress
* Updated sources to resolve some splint (static code analysis)
  issues
* Updates for OpenSDK (WinCE6/ARM) build target (patch contributed
  itsnotabigtruck)
* Added OSX Xcode3+ template files to Other Builds collection
  (contributed by Vasyl)

Ver 2.0.22 - Sat Sep 11 23:26:42 EDT 2010
* Fixed line width calculation for normal and textured polygons
  (patch contributed by Daniel (SupSuper) 
* Fixed version setup in configure.in
* Added rounded rectangle and rounded box primitives
* Updated test program for roundedbox + left/right-click support
* Fixed pie calculation (line-case detection, end-angle clipping)
* Updated DLL_EXPORT handling and added WIN64 test 
  (patch contributed by tigerfishdaisy)
* Added thick line primitive based on Murphy's algorithm

Ver 2.0.21 - Thu May 27 21:14:37 PDT 2010
* Add VC9 project files, source code formatting, fix compiler 
  warnings, move VC8 project files, remove leftover autoconf/m4 files
* Move static transfer lookup array from .h to .c 
  (patch contributed by Zbigniew Holdys)
* Add support for VC++ compiler/MASM assembly with active USE_MMX 
  (contributed by Markus Hossner)
* Added Haiku build support in configure.in 
  (contributed by Scott McCreary)
* Added helper function that multiplies the alpha channel in
  a 32bit surface (contributed by Zbigniew Holdys)
* Disable MMX for 64bit platforms due to lack of support for
  pusha/popa (contributed by Olivier Boudeville)
* Update all library functions for doxygen markup. Add doxygen
  configuration file.
* Added font rotation support and updated TestFonts sample
* Fixed rotate90 for 0 degree case
  (contributed by Chris Allport)
* Updated TestRotozoom with rotate90 case and message display
* Added support for SDL 1.3
* Added doxygen generated documentation.

Ver 2.0.20 - Wed Sep 23 18:42:02 PDT 2009
* Fix for 1-pixel postponement with 8bit scaling 
* Updates to headers and configure to allow for cross-compiling 
  to DLL (not just static .a) and fixes for compiling on Windows 
  using autotools
* Added Visual CE Project to Other Builds
* Added Symbian and Windows 64bit fix for lrint function.

Ver 2.0.19 - Mon Apr 20 22:56:59 PDT 2009

* gcc43 compiler patch for asm code in image filters 
  as per http://bugs.gentoo.org/219621 (Peter Alfredsen)
* off-by-one hline patch for 1byte indexed surfaces (Manuel Lausch)
* update acinclude.m4 from SDL-1.2 and moved # serial in file
* applied patch to add pkg-config support for SDL_gfx (Luca Bigliardi)
* changed configure.in INTERFACE_AGE and BINARY_AGE based on Debian
  build feedback (Barry deFreese)
* fixed zoom XY-flip code and made it available for all modes in
  zoom and rotations
* fixed edge issue on interpolated rotozoom modes
* updated TestRotozoom custom test mode, added delay parameter, added
  tests which include flipping, included negative rotation
* update libtool files
* updated Makefile.mingw (Jan Leike)

Ver 2.0.18 - Sun Dec 21 08:38:20 PST 2008

* libtool update for dependencies
* MSVC C89 variable hoisting to enable MSVC compile (sweetlilmre)
* addition of VC9 (Visual Studio 2008) project file (sweetlilmre)
* update to MinGW make file (Sergio Padrino, sweetlilmre)
* fix uninitialised variables in _putPixelAlpha() (Paul Pedriana)
* reverted to original alpha routine in _putPixelAlpha() as new one was causing issues in 16bbp
  new routine #define has been renamed to EXPERIMENTAL_ALPHA_PIXEL_ROUTINE
  original is now DEFAULT_ALPHA_PIXEL_ROUTINE
* added colorkey handling patch to rotozoom (Sergio Padrino)
* downgraded optimization to -O for asm compatibility with gcc4


Ver 2.0.17 - Sun Jun  1 15:39:38 EDT 2008

* hline/vline clipping and swap-logic fixed
* clipping box dimension check added and clipping-checks optimized
* TestFonts program updated
* Updated documentation on cross-compilation, AMD64
* Fixed circleColor overdraw
* Added arcColor routine
* Added polygonColorMT and polygonTextureMT routines
* Updated 32bit alpha blending routine


Ver 2.0.16 - Mon Feb  5 19:03:31 AST 2007
* updated config.sub to newer version 
* added custom RGBA blitter function that sets/maintains target alpha
* added setAlpha function that sets alpha values in 32bit surfaces
* added TextGfxBlit testprogram for new functions
* Fixed C++ type in .h files (thanks Olivier)
* Update some text in README and .spec file
* Removed broken Uint filter routines from test program
* "Cosmetic" changes to test programs and Automake.am


Ver 2.0.15 - Fri Dec 22 08:44:31 AST 2006
* bugfixes in 32bit _putPixelAlpha/filledRectAlpha (thanks CISC)
* various bugfixes in return value generation and return (thanks CISC)
* lock font-surface before using it (thanks CISC)
* fix some breakage in non-MMX filter routines (thanks CISC)
* fix typo in TestABGR (thanks CISC)
* fix double buffer flipping in Tets programs (thanks CISC)
* fixed microversion in .h
* fixed texturePolygon inverted bug
* fixed wrong surface use in test 21 of TestRotozoom
* added pitch handling in font surface (thanks CISC)


Ver 2.0.14 - Tue Dec 19 08:49:02 AST 2006
* added texturedPolygon routine to library (thanks Jees)
* changed iterator condition in aaellipse to add some overdraw to smooth circle
* added ShrinkImage routine to rotozoom code
* added TestABGR Test program (bug in SDL_gfx or SDL??)
* improved TestImageFilter program (bug in Uint filters) 
* changed .so versioning from .so.13.0.0 to .so.0.0.14
  (best is to remove all old libs before installing)
* added updated MacOSX ProjectBuilder file (thanks Brian)
* added info on MacOSX 10.3 build
* added mingw (on cygwin) makefile (thanks Brian)
* added Dev-Cpp makefile (thanks Sebastian)
* removed some unused variables from gfx and rotozoom code (thanks Thi)
* fixed a non-critical typecast mistake (thanks Thi)
* converted all C++ comments to C comments in SDL_imageFilter


Ver 2.0.13 - Tue Dec 21 08:41:25 EST 2004
* changed include back to "SDL.h"
* compile fixes for OSX fink  (thanks (Michael)

SDL_gfxBlitFunc.c
/* 

SDL_gfxBlitFunc: custom blitters (part of SDL_gfx library)

LGPL (c) A. Schiffler

*/

#include "SDL_gfxBlitFunc.h"

/*!  
\brief Alpha adjustment table for custom blitter.

The table provides values for a modified, non-linear 
transfer function which maintain brightness.

*/
static unsigned int GFX_ALPHA_ADJUST_ARRAY[256] = {
	0,  /* 0 */
	15,  /* 1 */
	22,  /* 2 */
	27,  /* 3 */
	31,  /* 4 */
	35,  /* 5 */
	39,  /* 6 */
	42,  /* 7 */
	45,  /* 8 */
	47,  /* 9 */
	50,  /* 10 */
	52,  /* 11 */
	55,  /* 12 */
	57,  /* 13 */
	59,  /* 14 */
	61,  /* 15 */
	63,  /* 16 */
	65,  /* 17 */
	67,  /* 18 */
	69,  /* 19 */
	71,  /* 20 */
	73,  /* 21 */
	74,  /* 22 */
	76,  /* 23 */
	78,  /* 24 */
	79,  /* 25 */
	81,  /* 26 */
	82,  /* 27 */
	84,  /* 28 */
	85,  /* 29 */
	87,  /* 30 */
	88,  /* 31 */
	90,  /* 32 */
	91,  /* 33 */
	93,  /* 34 */
	94,  /* 35 */
	95,  /* 36 */
	97,  /* 37 */
	98,  /* 38 */
	99,  /* 39 */
	100,  /* 40 */
	102,  /* 41 */
	103,  /* 42 */
	104,  /* 43 */
	105,  /* 44 */
	107,  /* 45 */
	108,  /* 46 */
	109,  /* 47 */
	110,  /* 48 */
	111,  /* 49 */
	112,  /* 50 */
	114,  /* 51 */
	115,  /* 52 */
	116,  /* 53 */
	117,  /* 54 */
	118,  /* 55 */
	119,  /* 56 */
	120,  /* 57 */
	121,  /* 58 */
	122,  /* 59 */
	123,  /* 60 */
	124,  /* 61 */
	125,  /* 62 */
	126,  /* 63 */
	127,  /* 64 */
	128,  /* 65 */
	129,  /* 66 */
	130,  /* 67 */
	131,  /* 68 */
	132,  /* 69 */
	133,  /* 70 */
	134,  /* 71 */
	135,  /* 72 */
	136,  /* 73 */
	137,  /* 74 */
	138,  /* 75 */
	139,  /* 76 */
	140,  /* 77 */
	141,  /* 78 */
	141,  /* 79 */
	142,  /* 80 */
	143,  /* 81 */
	144,  /* 82 */
	145,  /* 83 */
	146,  /* 84 */
	147,  /* 85 */
	148,  /* 86 */
	148,  /* 87 */
	149,  /* 88 */
	150,  /* 89 */
	151,  /* 90 */
	152,  /* 91 */
	153,  /* 92 */
	153,  /* 93 */
	154,  /* 94 */
	155,  /* 95 */
	156,  /* 96 */
	157,  /* 97 */
	158,  /* 98 */
	158,  /* 99 */
	159,  /* 100 */
	160,  /* 101 */
	161,  /* 102 */
	162,  /* 103 */
	162,  /* 104 */
	163,  /* 105 */
	164,  /* 106 */
	165,  /* 107 */
	165,  /* 108 */
	166,  /* 109 */
	167,  /* 110 */
	168,  /* 111 */
	168,  /* 112 */
	169,  /* 113 */
	170,  /* 114 */
	171,  /* 115 */
	171,  /* 116 */
	172,  /* 117 */
	173,  /* 118 */
	174,  /* 119 */
	174,  /* 120 */
	175,  /* 121 */
	176,  /* 122 */
	177,  /* 123 */
	177,  /* 124 */
	178,  /* 125 */
	179,  /* 126 */
	179,  /* 127 */
	180,  /* 128 */
	181,  /* 129 */
	182,  /* 130 */
	182,  /* 131 */
	183,  /* 132 */
	184,  /* 133 */
	184,  /* 134 */
	185,  /* 135 */
	186,  /* 136 */
	186,  /* 137 */
	187,  /* 138 */
	188,  /* 139 */
	188,  /* 140 */
	189,  /* 141 */
	190,  /* 142 */
	190,  /* 143 */
	191,  /* 144 */
	192,  /* 145 */
	192,  /* 146 */
	193,  /* 147 */
	194,  /* 148 */
	194,  /* 149 */
	195,  /* 150 */
	196,  /* 151 */
	196,  /* 152 */
	197,  /* 153 */
	198,  /* 154 */
	198,  /* 155 */
	199,  /* 156 */
	200,  /* 157 */
	200,  /* 158 */
	201,  /* 159 */
	201,  /* 160 */
	202,  /* 161 */
	203,  /* 162 */
	203,  /* 163 */
	204,  /* 164 */
	205,  /* 165 */
	205,  /* 166 */
	206,  /* 167 */
	206,  /* 168 */
	207,  /* 169 */
	208,  /* 170 */
	208,  /* 171 */
	209,  /* 172 */
	210,  /* 173 */
	210,  /* 174 */
	211,  /* 175 */
	211,  /* 176 */
	212,  /* 177 */
	213,  /* 178 */
	213,  /* 179 */
	214,  /* 180 */
	214,  /* 181 */
	215,  /* 182 */
	216,  /* 183 */
	216,  /* 184 */
	217,  /* 185 */
	217,  /* 186 */
	218,  /* 187 */
	218,  /* 188 */
	219,  /* 189 */
	220,  /* 190 */
	220,  /* 191 */
	221,  /* 192 */
	221,  /* 193 */
	222,  /* 194 */
	222,  /* 195 */
	223,  /* 196 */
	224,  /* 197 */
	224,  /* 198 */
	225,  /* 199 */
	225,  /* 200 */
	226,  /* 201 */
	226,  /* 202 */
	227,  /* 203 */
	228,  /* 204 */
	228,  /* 205 */
	229,  /* 206 */
	229,  /* 207 */
	230,  /* 208 */
	230,  /* 209 */
	231,  /* 210 */
	231,  /* 211 */
	232,  /* 212 */
	233,  /* 213 */
	233,  /* 214 */
	234,  /* 215 */
	234,  /* 216 */
	235,  /* 217 */
	235,  /* 218 */
	236,  /* 219 */
	236,  /* 220 */
	237,  /* 221 */
	237,  /* 222 */
	238,  /* 223 */
	238,  /* 224 */
	239,  /* 225 */
	240,  /* 226 */
	240,  /* 227 */
	241,  /* 228 */
	241,  /* 229 */
	242,  /* 230 */
	242,  /* 231 */
	243,  /* 232 */
	243,  /* 233 */
	244,  /* 234 */
	244,  /* 235 */
	245,  /* 236 */
	245,  /* 237 */
	246,  /* 238 */
	246,  /* 239 */
	247,  /* 240 */
	247,  /* 241 */
	248,  /* 242 */
	248,  /* 243 */
	249,  /* 244 */
	249,  /* 245 */
	250,  /* 246 */
	250,  /* 247 */
	251,  /* 248 */
	251,  /* 249 */
	252,  /* 250 */
	252,  /* 251 */
	253,  /* 252 */
	253,  /* 253 */
	254,  /* 254 */
	255   /* 255 */
};

/*!
\brief Internal blitter using adjusted destination alpha during RGBA->RGBA blits.

Performs the blit based on the 'info' structure and applies the transfer function
to the destination 'a' values.

\param info The blit info to use.
*/
void _SDL_gfxBlitBlitterRGBA(SDL_gfxBlitInfo * info)
{
	int       width = info->d_width;
	int       height = info->d_height;
	Uint8    *src = info->s_pixels;
	int       srcskip = info->s_skip;
	Uint8    *dst = info->d_pixels;
	int       dstskip = info->d_skip;
	SDL_PixelFormat *srcfmt = info->src;
	SDL_PixelFormat *dstfmt = info->dst;
	Uint8       srcbpp = srcfmt->BytesPerPixel;
	Uint8       dstbpp = dstfmt->BytesPerPixel;

	while (height--) {
		GFX_DUFFS_LOOP4( {
			Uint32 pixel;
			unsigned sR;
			unsigned sG;
			unsigned sB;
			unsigned sA;
			unsigned dR;
			unsigned dG;
			unsigned dB;
			unsigned dA;
			unsigned sAA;
			GFX_DISASSEMBLE_RGBA(src, srcbpp, srcfmt, pixel, sR, sG, sB, sA);
			GFX_DISASSEMBLE_RGBA(dst, dstbpp, dstfmt, pixel, dR, dG, dB, dA);
			sAA=GFX_ALPHA_ADJUST_ARRAY[sA & 255];
			GFX_ALPHA_BLEND(sR, sG, sB, sAA, dR, dG, dB);
			dA |= sAA;
			GFX_ASSEMBLE_RGBA(dst, dstbpp, dstfmt, dR, dG, dB, dA);
			src += srcbpp; dst += dstbpp;
		}, width);
		src += srcskip;
		dst += dstskip;
	}
}

/*!
\brief Internal blitter setup wrapper for RGBA->RGBA blits.

Sets up the blitter info based on the 'src' and 'dst' surfaces and rectangles.

\param src The source surface.
\param srcrect The source rectangle.
\param dst The destination surface.
\param dstrect The destination rectangle.

\returns Returns 1 if blit was performed, 0 otherwise.
*/
int _SDL_gfxBlitRGBACall(SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst, SDL_Rect * dstrect)
{
	/*
	* Set up source and destination buffer pointers, then blit 
	*/
	if (srcrect->w && srcrect->h) {
		SDL_gfxBlitInfo info;

		/*
		* Set up the blit information 
		*/
#if (SDL_MINOR_VERSION == 3)
		info.s_pixels = (Uint8 *) src->pixels               + (Uint16) srcrect->y * src->pitch + (Uint16) srcrect->x * src->format->BytesPerPixel;
#else
		info.s_pixels = (Uint8 *) src->pixels + src->offset + (Uint16) srcrect->y * src->pitch + (Uint16) srcrect->x * src->format->BytesPerPixel;
#endif
		info.s_width = srcrect->w;
		info.s_height = srcrect->h;
		info.s_skip = (int)(src->pitch - info.s_width * src->format->BytesPerPixel);
#if (SDL_MINOR_VERSION == 3)
		info.d_pixels = (Uint8 *) dst->pixels               + (Uint16) dstrect->y * dst->pitch + (Uint16) dstrect->x * dst->format->BytesPerPixel;
#else
		info.d_pixels = (Uint8 *) dst->pixels + dst->offset + (Uint16) dstrect->y * dst->pitch + (Uint16) dstrect->x * dst->format->BytesPerPixel;
#endif
		info.d_width = dstrect->w;
		info.d_height = dstrect->h;
		info.d_skip = (int)(dst->pitch - info.d_width * dst->format->BytesPerPixel);
		info.aux_data = NULL;
		info.src = src->format;
		info.table = NULL;
		info.dst = dst->format;

		/*
		* Run the actual software blitter 
		*/
		_SDL_gfxBlitBlitterRGBA(&info);
		return 1;
	}

	return (0);
}

/*!
\brief Blitter for RGBA->RGBA blits with alpha adjustment.

Verifies the input 'src' and 'dst' surfaces and rectangles and performs blit.
The destination clip rectangle is honored.

\param src The source surface.
\param srcrect The source rectangle.
\param dst The destination surface.
\param dstrect The destination rectangle.

\returns Returns 1 if blit was performed, 0 otherwise, or -1 if an error occured.
*/
int SDL_gfxBlitRGBA(SDL_Surface * src, SDL_Rect * srcrect, SDL_Surface * dst, SDL_Rect * dstrect)
{
	SDL_Rect  sr, dr;
	int       srcx, srcy, w, h;

	/*
	* Make sure the surfaces aren't locked 
	*/
	if (!src || !dst) {
		SDL_SetError("SDL_UpperBlit: passed a NULL surface");
		return (-1);
	}
	if ((src->locked) || (dst->locked)) {
		SDL_SetError("Surfaces must not be locked during blit");
		return (-1);
	}

	/*
	* If the destination rectangle is NULL, use the entire dest surface 
	*/
	if (dstrect == NULL) {
		dr.x = dr.y = 0;
		dr.w = dst->w;
		dr.h = dst->h;
	} else {
		dr = *dstrect;
	}

	/*
	* Clip the source rectangle to the source surface 
	*/
	if (srcrect) {
		int       maxw, maxh;

		srcx = srcrect->x;
		w = srcrect->w;
		if (srcx < 0) {
			w += srcx;
			dr.x -= srcx;
			srcx = 0;
		}
		maxw = src->w - srcx;
		if (maxw < w)
			w = maxw;

		srcy = srcrect->y;
		h = srcrect->h;
		if (srcy < 0) {
			h += srcy;
			dr.y -= srcy;
			srcy = 0;
		}
		maxh = src->h - srcy;
		if (maxh < h)
			h = maxh;

	} else {
		srcx = srcy = 0;
		w = src->w;
		h = src->h;
	}

	/*
	* Clip the destination rectangle against the clip rectangle 
	*/
	{
		SDL_Rect *clip = &dst->clip_rect;
		int       dx, dy;

		dx = clip->x - dr.x;
		if (dx > 0) {
			w -= dx;
			dr.x += dx;
			srcx += dx;
		}
		dx = dr.x + w - clip->x - clip->w;
		if (dx > 0)
			w -= dx;

		dy = clip->y - dr.y;
		if (dy > 0) {
			h -= dy;
			dr.y += dy;
			srcy += dy;
		}
		dy = dr.y + h - clip->y - clip->h;
		if (dy > 0)
			h -= dy;
	}

	if (w > 0 && h > 0) {
		sr.x = srcx;
		sr.y = srcy;
		sr.w = dr.w = w;
		sr.h = dr.h = h;
		return (_SDL_gfxBlitRGBACall(src, &sr, dst, &dr));
	}

	return 0;
}

/*!
\brief Sets the alpha channel in a 32 bit surface.

Helper function that sets the alpha channel in a 32 bit surface
to a constant value.
Only 32 bit surfaces can be used with this function.

\param src Pointer to the target surface to change.
\param a The alpha value to set.

\return Returns 1 if alpha was changed, -1 otherwise.
*/
int SDL_gfxSetAlpha(SDL_Surface *src, Uint8 a)
{
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
	const int alpha_offset = 0;
#else
	const int alpha_offset = 3;
#endif
	int i, j, row_skip;
	Uint8 *pixels;
	
	/* Check if we have a 32bit surface */
	if ( (src==NULL) || 
	     (src->format==NULL) || 
	     (src->format->BytesPerPixel!=4) ) {
          SDL_SetError("SDL_gfxSetAlpha: Invalid input surface.");
	  return -1;
        }

	/*
	* Lock the surface 
	*/
	if (SDL_MUSTLOCK(src)) {
		if (SDL_LockSurface(src) < 0) {
			return (-1);
		}
	}
        
        /* Process */
        pixels = (Uint8 *)src->pixels;
        row_skip = (src->pitch - (4*src->w));
        pixels += alpha_offset;
        for ( i=0; i<src->h; i++ ) {
          for ( j=0; j<src->w; j++  ) {
            *pixels = a; 
            pixels += 4;
          }
          pixels += row_skip;
        }

	/*
	* Unlock surface 
	*/
	if (SDL_MUSTLOCK(src)) {
		SDL_UnlockSurface(src);
	}
        
        return 1; 
}

/*!
\brief Multiply the alpha channel in a 32bit surface.

Helper function that multiplies the alpha channel in a 32 bit surface
with a constant value. The final alpha is always scaled to the range 
0-255 (i.e. the factor is a/256).

Only 32 bit surfaces can be used with this function.

\param src Pointer to the target surface to change.
\param a The alpha value to multiply with. When a is 255, this function is a NoOp.

\return Returns 1 if alpha was changed, 0 otherwise. Returns -1 if input surface is invalid.
*/
int SDL_gfxMultiplyAlpha(SDL_Surface *src, Uint8 a)
{
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
	const int alpha_offset = 0;
#else
	const int alpha_offset = 3;
#endif
	int i, j, row_skip;
	Uint8 *pixels;
        
	/* Check if we have a 32bit surface */
	if ( (src==NULL) || 
	     (src->format==NULL) || 
	     (src->format->BytesPerPixel!=4) ) {
          SDL_SetError("SDL_gfxMultiplyAlpha: Invalid input surface.");
	  return -1;
        }
        
        /* Check if multiplication is needed */
        if (a==255) {
          return 0;
        }

	/*
	* Lock the surface 
	*/
	if (SDL_MUSTLOCK(src)) {
		if (SDL_LockSurface(src) < 0) {
			return (-1);
		}
	}
        
        /* Process */
        pixels = (Uint8 *)src->pixels;
        row_skip = (src->pitch - (4*src->w));
        pixels += alpha_offset;
        for ( i=0; i<src->h; i++ ) {
  	  for ( j=0; j<src->w; j++  ) {
            *pixels = (Uint8)(((int)(*pixels)*a)>>8);
            pixels += 4;
          }
          pixels += row_skip;
        }

	/*
	* Unlock surface 
	*/
	if (SDL_MUSTLOCK(src)) {
		SDL_UnlockSurface(src);
	}
		
        return 1;
}
